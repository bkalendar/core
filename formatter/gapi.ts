import { icalRrule } from "@/formatter/ical.ts";
import gapi from "gapi.calendar";
import { Timetable } from "@/timetable.ts";
import { ASIA_HO_CHI_MINH, dateOfIndex } from "@/formatter/utils.ts";

type EventInput = gapi.client.calendar.EventInput;

export function formatGapi(timetable: Required<Timetable>): EventInput[] {
	const events = [];
	for (const timerow of timetable.rows) {
		if (isNaN(timerow.weekday) || timerow.weeks.length == 0) {
			continue;
		}

		const start = timerow.weeks.findIndex(Boolean);
		const recurrence = icalRrule(timerow, timetable.startMondayUTC);
		const event: EventInput = {
			summary: timerow.name,
			description: Object.entries(timerow.extras).map((e) => e.join(": ")).join("\n"),
			location: timerow.location,
			start: {
				dateTime: dateOfIndex(start, timerow.startHm, timetable.startMondayUTC, timerow.weekday)
					.toISOString(),
				timeZone: ASIA_HO_CHI_MINH,
			},
			end: {
				dateTime: dateOfIndex(start, timerow.endHm, timetable.startMondayUTC, timerow.weekday)
					.toISOString(),
				timeZone: ASIA_HO_CHI_MINH,
			},
			recurrence,
		};
		events.push(event);
	}
	return events;
}
