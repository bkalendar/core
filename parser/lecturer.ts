import {
	NumOfColumnMismatchError,
	ParseError,
	SemesterNotFoundError,
	TableNotFoundError,
} from "@/parser/errors.ts";
import { Timerow, Timetable } from "@/timetable.ts";
import { parseTime } from "@/parser/utils.ts";

const HEADER = "Lớp\tTên MH\tPhòng\tDãy\tThứ\tSố tiết\tTiết\tGiờ\tTuần học\t% ND";
const COLS = HEADER.split("\t").length;

export function parse(src: string): Timetable {
	src = src.trim();

	let matches = src.match(/Năm học (\d+)/);
	if (!matches) {
		throw new SemesterNotFoundError(src);
	}
	const yearFrom = Number(matches[1]);

	matches = src.match(/Học kỳ ([123])/);
	if (!matches) {
		throw new SemesterNotFoundError(src);
	}
	const semester = (yearFrom % 100) * 10 + Number(matches[1]);

	const lines = src.split("\n");
	for (let i = 0; i < lines.length; i++) {
		if (lines[i] == HEADER) {
			return {
				semester,
				rows: parseFromHeader(lines, i),
			};
		}
	}
	throw new TableNotFoundError(src);
}

function parseFromHeader(lines: string[], i: number): Timerow[] {
	const rows = [];
	i++;
	for (; i < lines.length; i++) {
		if (lines[i].startsWith("Đang xem")) break;

		// a sample timerow:
		//       (0)                    (1)             (2)   (3)(4)(5)(6)      (7)                   (8)              (9)
		// 20221_CO1006_L11..NHAP MON DIEN TOAN (TH)..H6-707..H6..5.. .. ..12:00 - 16:50..--|--|--|--|--|--|--|--|43|..0%
		const cols = lines[i].split("\t");
		if (cols.length != COLS) {
			throw new NumOfColumnMismatchError(lines[i], COLS, cols.length);
		}
		const weekday = Number(cols[4]);
		const times = cols[7].split(" - ");
		if (times.length != 2) {
			throw new ParseError(cols[7], `Time column is not separated by " - "`);
		}
		const startHm = parseTime(times[0]);
		const endHm = parseTime(times[1]);
		// trim the trailing '|'
		const weeks = cols[8].slice(0, -1).split("|").map((s) => s == "--" ? null : Number(s));

		const row = {
			name: cols[1],
			weekday,
			startHm,
			endHm,
			weeks,
			location: cols[2],
			extras: {
				"lớp": cols[0],
			},
		} satisfies Timerow;

		rows.push(row);
	}

	return rows;
}
