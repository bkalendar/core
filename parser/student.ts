import {
	NumOfColumnMismatchError,
	ParseError,
	SemesterNotFoundError,
	TableNotFoundError,
} from "@/parser/errors.ts";
import { Timerow, Timetable } from "@/timetable.ts";
import { parseTime } from "@/parser/utils.ts";

const HEADER =
	"MÃ MH\tTÊN MÔN HỌC\tTÍN CHỈ\tTC HỌC PHÍ\tNHÓM-TỔ\tTHỨ\tTIẾT\tGIỜ HỌC\tPHÒNG\tCƠ SỞ\tTUẦN HỌC";
const COLS = HEADER.split("\t").length;

export function parseStudent(src: string): Timetable {
	src = src.trim();
	const lines = src.split("\n");
	for (let i = 0; i < lines.length; i++) {
		if (lines[i] == HEADER) {
			return parseFromHeader(lines, i);
		}
	}
	throw new TableNotFoundError(src);
}

function parseFromHeader(lines: string[], i: number): Timetable {
	// from the header, go back 2 lines and extract the semester
	const matches = lines[i - 2]?.match(/^Học kỳ ([123]) Năm học (\d+) - (\d+)$/);
	if (!matches) {
		throw new SemesterNotFoundError(lines[i - 2]);
	}
	// "Học kỳ 2 năm học 2023 - ..." should yield semester=232
	const semester = (Number(matches[2]) % 100) * 10 + Number(matches[1]);

	const rows = [];
	i++;
	for (; i < lines.length; i++) {
		if (lines[i].startsWith("Tổng số tín chỉ đăng ký:")) break;

		// a sample timerow:
		//   (0)       (1)       (2)(3) (4) (5)  (6)        (7)        (8)     (9)                     (10)
		// CH1003..Hóa đại cương..3..3..L14..6..10-12..15:00 - 17:50..H2-301..BK-CS2..--|09|10|11|12|13|14|15|16|--|18|19|
		const cols = lines[i].split("\t");
		if (cols.length != COLS) {
			throw new NumOfColumnMismatchError(lines[i], COLS, cols.length);
		}
		const weekday = Number(cols[5]);
		const times = cols[7].split(" - ");
		if (times.length != 2) {
			throw new ParseError(cols[7], `Time column is not separated by " - "`);
		}
		const startHm = parseTime(times[0]);
		const endHm = parseTime(times[1]);
		// trim the trailing '|'
		const weeks = cols[10].slice(0, -1).split("|").map((s) => s == "--" ? null : Number(s));

		const row = {
			name: cols[1],
			weekday,
			startHm,
			endHm,
			weeks,
			location: cols[8],
			extras: {
				"mã môn học": cols[0],
				"tín chỉ": cols[2],
				"tín chỉ học phí": cols[3],
				"nhóm tổ": cols[4],
			},
		} satisfies Timerow;

		rows.push(row);
	}

	return { semester, rows };
}
