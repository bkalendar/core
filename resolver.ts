import type { Timerow, Timetable } from "@/timetable.ts";
import { DAY, WEEK } from "@std/datetime";

export function resolve(timetable: Timetable): asserts timetable is Required<Timetable> {
	// already resolved
	if (timetable.startMondayUTC) return;

	for (const timerow of timetable.rows) {
		if (isUnresolvable(timerow)) {
			continue;
		}
		const { indexOfWeek1, weekOfIndex0 } = findIndex0AndWeek1(timerow.weeks);
		const mondayOfIndex0 = findMondayOfIndex0(indexOfWeek1, weekOfIndex0, timetable.semester);
		timetable.startMondayUTC ??= mondayOfIndex0;
		if (+timetable.startMondayUTC != +mondayOfIndex0) {
			throw new MixedSemesterError(timetable.startMondayUTC, mondayOfIndex0);
		}
	}

	// if calendar contains no resolvable timerow
	if (!timetable.startMondayUTC) {
		throw new UnresolvedError();
	}
}

export function isUnresolvable(timerow: Timerow): boolean {
	return isNaN(timerow.weekday) || timerow.weeks.length == 0 ||
		(timerow.weeks.length == 1 && timerow.weeks[0] == null);
}

export class MixedSemesterError extends Error {
	constructor(public doiz1: Date, public doiz2: Date) {
		super(
			`Semester ambiguity: Two possible semester start dates: ${doiz1.toISOString()} and ${doiz2.toISOString()}`,
		);
	}
}

export class UnresolvedError extends Error {
	constructor() {
		super(`Calendar cannot be resolved`);
	}
}

/**
 * Find week 1 and index 0 for semester to year conversion.
 *
 * For example:
 *
 * ```
 * --|--|--|--|02|--|--|--
 *          ^^ we know week 01 should be here, so indexOfWeek1=3
 * ^^ we don't know the week at index 0
 * ```
 *
 * ```
 * --|33|--|--|--|--|38|--
 * ^^ we know week 32 should be at index 0, so weekOfIndex0=32
 * ```
 *
 * ```
 * --|--|49|--|--|--|--|--|--|03
 * ^^ weekOfIndex0=47   ^^ indexOfWeek1=7
 * ```
 */
function findIndex0AndWeek1(weeks: (number | null)[]) {
	let weekOfIndex0: number | undefined;
	let indexOfWeek1: number | undefined;

	for (const [index, week] of weeks.entries()) {
		if (week == null) continue;

		// for example, week 2 at index 2 means new year's at index 1
		if (week - index <= 0) {
			indexOfWeek1 = index - week + 1;
			break;
		}

		// else, initialize base week
		if (!weekOfIndex0) {
			weekOfIndex0 = week - index;
		}
	}
	return { weekOfIndex0, indexOfWeek1 };
}

function findMondayOfIndex0(
	indexOfWeek1: number | undefined,
	weekOfIndex0: number | undefined,
	semester: number,
) {
	const year = 2000 + Math.trunc(semester / 10);

	if (indexOfWeek1) { // has a new year break, hence it span 2 years
		// start with new year
		let d = +Date.UTC(year + 1, 0, 4);
		// offset new year into index 0 week
		d -= indexOfWeek1 * WEEK;
		// offset weekday
		d -= (dayOfWeek(d) - 2) * DAY;

		return new Date(d);
	}

	switch (semester % 10) {
		case 1: // no new year break, semester 1 always starts early
			return index0Starts("early");
		case 2: // no new year break, semester 2 MAY start early
			return index0Starts(weekOfIndex0! >= 26 ? "early" : "late");
		case 3: // no new year break, semester 3 always starts late
			return index0Starts("late");
		default:
			throw new Error("unreachable");
	}

	function index0Starts(type: "early" | "late") {
		// start with new year
		let d = +Date.UTC(type == "early" ? year : year + 1, 0, 4);
		// offset new year into index 0 week
		d += (weekOfIndex0! - 1) * WEEK;
		// offset weekday
		d -= (dayOfWeek(d) - 2) * DAY;

		return new Date(d);
	}
}

function dayOfWeek(d: number): number {
	return (new Date(d).getDay() + 6) % 7 + 2;
}
