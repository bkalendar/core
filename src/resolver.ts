import { resolveFirstDate } from './date_utils';
import type { EntryRaw, Entry } from './entry';
import type { TimetableRaw, Timetable, parseTimetables } from './timetable';
import type { SemesterContext } from './date_utils';

/**
 * Resolve raw timetables into resolved timetables.
 *
 * This takes care of "orphans" and merges entries.
 *
 * @param timetableRaws list of raw timetables,
 * preferably from {@link parseTimetables}
 */
export function resolveTimetables(
    timetableRaws: TimetableRaw[]
): Timetable[] {
    const newTimetables: Timetable[] = [];
    let allOrphans: EntryRaw[] = [];

    for (const timetable of timetableRaws) {
        let firstDate = commonFirstDate(timetable);
        const { resolved, orphans } = resolveEntries(timetable, firstDate);
        newTimetables.push({
            ...timetable,
            entries: resolved,
            start: firstDate,
        });
        allOrphans.push(...orphans);
    }
    for (const timetable of newTimetables) {
        let rawTimetable: TimetableRaw = { ...timetable, entries: allOrphans };
        const { resolved, orphans } = resolveEntries(
            rawTimetable,
            timetable.start
        );
        allOrphans = orphans;
        timetable.entries.push(...resolved);
        timetable.entries.sort((a, b) => {
            let diff = a.wday - b.wday;
            if (diff === 0) diff = a.start - b.start;
            return diff;
        });
        mergeEntriesResolved(timetable.entries);
    }

    if (allOrphans.length != 0) {
        console.warn("Found orphan(s):", allOrphans);
    }

    return newTimetables;
}

function commonFirstDate(timetable: TimetableRaw): Date {
    const multiset: Record<number, number> = {};
    for (const entry of timetable.entries) {
        // we can do nothing with empty weeks
        if (entry.weeks.find(Boolean) === undefined) continue;
        const firstDate = resolveFirstDate(entry.weeks, timetable);
        // get or init 0
        const currentCount = multiset[+firstDate] || 0;
        multiset[+firstDate] = currentCount + 1;
    }
    // NOTE: potentially empty multiset?
    const maxCount = Math.max(...Object.values(multiset));
    const maxKey = Object.entries(multiset).find(
        ([_, count]) => count == maxCount
    )[0];
    return new Date(Number(maxKey));
}


/**
 * Resolve raw entries using a start date.
 *
 * Rule out orphans (start at a different week than others)
 *
 * @param timetable raw timetable
 * @param start start date, from {@link SemesterContext}
 * @returns entries of that timetable and found orphans
 */
function resolveEntries(
    timetable: TimetableRaw,
    start: Date
): {
    resolved: Entry[];
    orphans: EntryRaw[];
} {
    const resolved: Entry[] = [];
    const orphans: EntryRaw[] = [];
    for (const entry of timetable.entries) {
        // empty weeks, huh?
        if (entry.weeks.find(Boolean) === undefined) {
            console.warn("Discard entry without weeks", entry);
            continue;
        }
        const firstDate = resolveFirstDate(entry.weeks, timetable);
        if (+start === +firstDate) {
            const firstWeek = entry.weeks.findIndex(Boolean);
            const lastWeek = entry.weeks.length - 1;
            resolved.push({
                ...entry,
                firstWeek,
                lastWeek,
                excludeWeeks: entry.weeks
                    // get indices of only falsy weeks
                    .map((n, i) => isNaN(n) && i),
                // slice to make firstWeek first element
                // .slice(firstWeek)
                // filter out truthy weeks
                // .filter(Boolean),
                // NOTE: above logic moved to entry merging process
            });
        } else {
            orphans.push(entry);
        }
    }
    return { resolved, orphans };
}

/**
 * Merge two entries. Throw if unidentical, so remember to check the hash before merging.
 * @returns Merged entry
 */
function mergeEntryResolved(
    entry1: Entry,
    entry2: Entry
): Entry {
    if (entry1.hash != entry2.hash)
        throw new Error("Cannot merge two different entries");

    let exclude1 = [...entry1.excludeWeeks, false];
    let exclude2 = [...entry2.excludeWeeks, false];
    for (let i = exclude1.length; i < exclude2.length; i += 1) {
        exclude1.push(i);
    }
    for (let i = exclude2.length; i < exclude1.length; i += 1) {
        exclude2.push(i);
    }
    // console.log(exclude1, exclude2);

    let sortedWeeks = [];
    for (
        let i = 0;
        i < Math.max(entry1.excludeWeeks.length, entry2.excludeWeeks.length);
        i += 1
    ) {
        if (exclude1[i] !== false && exclude2[i] !== false) sortedWeeks.push(i);
    }

    return {
        ...entry2,
        firstWeek: Math.min(entry1.firstWeek, entry2.firstWeek),
        lastWeek: Math.max(entry1.lastWeek, entry2.lastWeek),
        excludeWeeks: sortedWeeks,
    };
}

function mergeEntriesResolved(entries: Entry[]) {
    for (let i = 0; i < entries.length; i += 1) {
        if (entries[i].hash == entries[i + 1]?.hash) {
            entries[i] = mergeEntryResolved(entries[i], entries[i + 1]);
            entries.splice(i + 1, 1);
        }
        entries[i].excludeWeeks = entries[i].excludeWeeks.filter(
            (x: number | false) => x !== false
        );
    }
}