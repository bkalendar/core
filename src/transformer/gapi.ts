import { formatISO } from "date-fns";
import { rrule } from "./ical.js";
import type { MachineTimetable } from "./index.js";
import { transformInfoBasic } from "./info.js";

type EventInput = gapi.client.calendar.EventInput;

const TIME_ZONE = "Asia/Ho_Chi_Minh";

export function transform(
	{ timerows }: MachineTimetable,
	options?: { infoTransformer?: typeof transformInfoBasic }
): EventInput[] {
	let infoTransformer = options?.infoTransformer ?? transformInfoBasic;

	const events = [];
	for (const timerow of timerows) {
		if (!timerow.time) continue;
		const { start, end, until, exceptions } = timerow.time;
		const recurrence = rrule(until, exceptions);

		const { summary, description } = infoTransformer(timerow.info);

		const event: EventInput = {
			summary,
			description,
			location: timerow.location.room,
			start: {
				dateTime: formatISO(start),
				timeZone: TIME_ZONE,
			},
			end: {
				dateTime: formatISO(end),
				timeZone: TIME_ZONE,
			},
			recurrence,
		};
		events.push(event);
	}
	return events;
}
