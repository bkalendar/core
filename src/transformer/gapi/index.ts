import { formatISO } from "date-fns";
import type { Timetable } from "../../parser";
import { rrule } from "../ical";
import { transformTime } from "../time.js";

type EventInput = gapi.client.calendar.EventInput;

export const TIME_ZONE = "Asia/Ho_Chi_Minh";

export function transform({ timerows, semester }: Timetable): EventInput[] {
	const events = [];
	for (const timerow of timerows) {
		const transformed = transformTime(timerow.time, semester);
		if (transformed === null) continue;

		const { start, end, until, exceptions } = transformed;
		const recurrence = rrule(until, exceptions);

		const event: EventInput = {
			summary: `${timerow.info.course} - ${timerow.info.name}`,
			location: timerow.location.room,
			description: JSON.stringify(timerow.info),
			start: {
				dateTime: formatISO(start),
				timeZone: TIME_ZONE,
			},
			end: {
				dateTime: formatISO(end),
				timeZone: TIME_ZONE,
			},
			recurrence,
		};
		events.push(event);
	}
	return events;
}
