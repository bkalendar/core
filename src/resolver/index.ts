import type { Timerow, Timetable } from "../parser";
import { calcBase } from "../transformer/calcBase";
import { bestEntry, groupBy } from "../utils";

export function resolve(timetables: Timetable[]) {
	// sort timetables by ascending semester.
	// we do this to push all unresolved timerows into newer timetable
	// because unresolved timerows can't reference old timetable.
	timetables.sort((a, b) => {
		if (a.semester.year == b.semester.year) {
			return a.semester.semester - b.semester.semester;
		}
		return a.semester.year - b.semester.year;
	});

	// resolve the correct weeks of timetables, using majority votes
	let leftovers: Timerow[] = [];
	for (const timetable of timetables) {
		// pour all leftovers from previous loop to resolve
		timetable.timerows.push(...leftovers);
		leftovers = [];

		// standard groupBy operation
		const votes = groupBy(
			timetable.timerows,
			(timerow) =>
				timerow.time.weeks && calcBase(timerow.time.weeks, timetable.semester).getTime()
		);

		// store nullish entries, i.e, ones without weeks, "--|"
		const nullishs: Timerow[] = votes.get(null) || [];
		votes.delete(null);

		// picking best base with most votes
		const best = bestEntry(votes, (a, b) => a.length - b.length);

		// if we can't get a best base, welp that's it
		if (best === null) {
			console.error("there's no way i can resolve this timetable", timetable);
			leftovers.push(...timetable.timerows);
			continue;
		}

		// get voters result from votes
		const [bestKey, bestVoters] = best;
		const majority = bestVoters!;
		votes.delete(bestKey);
		const localLeftovers = Array.from(votes.values()).flat();

		timetable.timerows = majority.concat(nullishs);
		leftovers.push(...localLeftovers);
	}

	if (leftovers.length != 0) {
		console.error("unresolved timerows", leftovers);
	}
}
