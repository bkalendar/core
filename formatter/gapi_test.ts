import { parseStudent } from "@/parser/student.ts";
import { resolve } from "@/resolver.ts";
import { formatGapi } from "@/formatter/gapi.ts";
import { assertSnapshot } from "std/testing/snapshot.ts";
import { parseLecturer } from "@/parser/lecturer.ts";

Deno.test("snapshot student", async (t) => {
	const src = `
Học kỳ 1 Năm học 2020 - 2021
Ngày cập nhật:2021-01-14 13:44:46.0
MÃ MH	TÊN MÔN HỌC	TÍN CHỈ	TC HỌC PHÍ	NHÓM-TỔ	THỨ	TIẾT	GIỜ HỌC	PHÒNG	CƠ SỞ	TUẦN HỌC
MI1003	Giáo dục quốc phòng	--	1	L02	--	0-0	0:00 - 0:00	------	BK	--|--|--|--|--|--|45|46|47|48|
PE1023	Võ (Vovinam, Karate, Taewondo) (học phần 1)	--	1.5	L262	2	10-12	15:00 - 17:50	CS2-NHATHIDAU-SAN1	BK-DAn	--|--|--|42|43|44|--|--|--|--|49|50|--|52|53|01|02|03|
MT1003	Giải tích 1	4	4	L25	3	2-4	7:00 - 9:50	H1-304	BK-DAn	--|--|--|42|43|44|--|--|--|--|49|50|--|52|53|01|02|03|
PH1003	Vật lý 1	4	4	L24	4	8-10	13:00 - 15:50	H1-301	BK-DAn	--|--|--|42|43|44|--|--|--|--|49|50|--|52|53|01|02|03|
MT1004	GIAI TICH 1 (BT)	--	--	L44	5	10-11	15:00 - 16:50	H1-103	BK-DAn	--|--|--|42|43|44|--|--|--|--|49|50|--|52|53|01|02|03|
PE1023	Võ (Vovinam, Karate, Taewondo) (học phần 1)	--	1.5	L262	6	10-12	15:00 - 17:50	CS2-NHATHIDAU-SAN1	BK-DAn	--|--|--|42|43|44|--|--|--|--|49|50|--|52|--|01|02|03|
Tổng số tín chỉ đăng ký: 8
	`;
	const timetable = parseStudent(src);
	resolve(timetable);
	const gapi = formatGapi(timetable);
	await assertSnapshot(t, gapi);
});

Deno.test("snapshot lecturer", async (t) => {
	const src = `
Năm học 2022

Học kỳ 1
   
Tìm:
Lớp	Tên MH	Phòng	Dãy	Thứ	Số tiết	Tiết	Giờ	Tuần học	% ND
20221_CO1006_L11	NHAP MON DIEN TOAN (TH)	H6-707	H6	5			12:00 - 16:50	--|--|--|--|--|--|--|--|43|	0%
20221_CO1006_L11	NHAP MON DIEN TOAN (TH)	HANGOUT_TUONGTAC	LIVE_HOME	5			12:00 - 16:50	--|--|--|--|--|--|--|--|--|--|--|46|47|	
20221_CO1006_L23	NHAP MON DIEN TOAN (TH)	H6-707	H6	5			12:00 - 16:50	--|--|--|--|--|--|--|--|--|--|45|--|--|--|--|--|51|	0%
20221_CO1006_L23	NHAP MON DIEN TOAN (TH)	HANGOUT_TUONGTAC	LIVE_HOME	5			12:00 - 16:50	--|--|--|--|--|--|--|--|--|--|--|46|47|	
20221_CO1006_L11	NHAP MON DIEN TOAN (TH)	HANGOUT_TUONGTAC	LIVE_HOME	5			12:00 - 16:50	--|--|--|--|--|--|--|--|--|--|--|--|--|--|49|	
Đang xem 1 đến 5 trong tổng số 5 mục	`;
	const timetable = parseLecturer(src);
	resolve(timetable);
	const gapi = formatGapi(timetable);
	await assertSnapshot(t, gapi);
});
