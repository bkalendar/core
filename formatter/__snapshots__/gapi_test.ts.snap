export const snapshot = {};

snapshot[`snapshot student 1`] = `
[
  {
    description: "mã môn học: PE1023
tín chỉ: --
tín chỉ học phí: 1.5
nhóm tổ: L262",
    end: {
      dateTime: "2020-10-12T10:50:00.000Z",
      timeZone: "Asia/Ho_Chi_Minh",
    },
    location: "CS2-NHATHIDAU-SAN1",
    recurrence: [
      "RRULE:FREQ=WEEKLY;UNTIL=20210118T080000Z",
      "EXDATE;TZID=Asia/Ho_Chi_Minh:20201102T150000,20201109T150000,20201116T150000,20201123T150000,20201214T150000",
    ],
    start: {
      dateTime: "2020-10-12T08:00:00.000Z",
      timeZone: "Asia/Ho_Chi_Minh",
    },
    summary: "Võ (Vovinam, Karate, Taewondo) (học phần 1)",
  },
  {
    description: "mã môn học: MT1003
tín chỉ: 4
tín chỉ học phí: 4
nhóm tổ: L25",
    end: {
      dateTime: "2020-10-13T02:50:00.000Z",
      timeZone: "Asia/Ho_Chi_Minh",
    },
    location: "H1-304",
    recurrence: [
      "RRULE:FREQ=WEEKLY;UNTIL=20210119T000000Z",
      "EXDATE;TZID=Asia/Ho_Chi_Minh:20201103T070000,20201110T070000,20201117T070000,20201124T070000,20201215T070000",
    ],
    start: {
      dateTime: "2020-10-13T00:00:00.000Z",
      timeZone: "Asia/Ho_Chi_Minh",
    },
    summary: "Giải tích 1",
  },
  {
    description: "mã môn học: PH1003
tín chỉ: 4
tín chỉ học phí: 4
nhóm tổ: L24",
    end: {
      dateTime: "2020-10-14T08:50:00.000Z",
      timeZone: "Asia/Ho_Chi_Minh",
    },
    location: "H1-301",
    recurrence: [
      "RRULE:FREQ=WEEKLY;UNTIL=20210120T060000Z",
      "EXDATE;TZID=Asia/Ho_Chi_Minh:20201104T130000,20201111T130000,20201118T130000,20201125T130000,20201216T130000",
    ],
    start: {
      dateTime: "2020-10-14T06:00:00.000Z",
      timeZone: "Asia/Ho_Chi_Minh",
    },
    summary: "Vật lý 1",
  },
  {
    description: "mã môn học: MT1004
tín chỉ: --
tín chỉ học phí: --
nhóm tổ: L44",
    end: {
      dateTime: "2020-10-15T09:50:00.000Z",
      timeZone: "Asia/Ho_Chi_Minh",
    },
    location: "H1-103",
    recurrence: [
      "RRULE:FREQ=WEEKLY;UNTIL=20210121T080000Z",
      "EXDATE;TZID=Asia/Ho_Chi_Minh:20201105T150000,20201112T150000,20201119T150000,20201126T150000,20201217T150000",
    ],
    start: {
      dateTime: "2020-10-15T08:00:00.000Z",
      timeZone: "Asia/Ho_Chi_Minh",
    },
    summary: "GIAI TICH 1 (BT)",
  },
  {
    description: "mã môn học: PE1023
tín chỉ: --
tín chỉ học phí: 1.5
nhóm tổ: L262",
    end: {
      dateTime: "2020-10-16T10:50:00.000Z",
      timeZone: "Asia/Ho_Chi_Minh",
    },
    location: "CS2-NHATHIDAU-SAN1",
    recurrence: [
      "RRULE:FREQ=WEEKLY;UNTIL=20210122T080000Z",
      "EXDATE;TZID=Asia/Ho_Chi_Minh:20201106T150000,20201113T150000,20201120T150000,20201127T150000,20201218T150000,20210101T150000",
    ],
    start: {
      dateTime: "2020-10-16T08:00:00.000Z",
      timeZone: "Asia/Ho_Chi_Minh",
    },
    summary: "Võ (Vovinam, Karate, Taewondo) (học phần 1)",
  },
]
`;

snapshot[`snapshot lecturer 1`] = `
[
  {
    description: "lớp: 20221_CO1006_L11",
    end: {
      dateTime: "2022-10-27T09:50:00.000Z",
      timeZone: "Asia/Ho_Chi_Minh",
    },
    location: "H6-707",
    recurrence: [],
    start: {
      dateTime: "2022-10-27T05:00:00.000Z",
      timeZone: "Asia/Ho_Chi_Minh",
    },
    summary: "NHAP MON DIEN TOAN (TH)",
  },
  {
    description: "lớp: 20221_CO1006_L11",
    end: {
      dateTime: "2022-11-17T09:50:00.000Z",
      timeZone: "Asia/Ho_Chi_Minh",
    },
    location: "HANGOUT_TUONGTAC",
    recurrence: [
      "RRULE:FREQ=WEEKLY;UNTIL=20221124T050000Z",
    ],
    start: {
      dateTime: "2022-11-17T05:00:00.000Z",
      timeZone: "Asia/Ho_Chi_Minh",
    },
    summary: "NHAP MON DIEN TOAN (TH)",
  },
  {
    description: "lớp: 20221_CO1006_L23",
    end: {
      dateTime: "2022-11-10T09:50:00.000Z",
      timeZone: "Asia/Ho_Chi_Minh",
    },
    location: "H6-707",
    recurrence: [
      "RRULE:FREQ=WEEKLY;UNTIL=20221222T050000Z",
      "EXDATE;TZID=Asia/Ho_Chi_Minh:20221117T120000,20221124T120000,20221201T120000,20221208T120000,20221215T120000",
    ],
    start: {
      dateTime: "2022-11-10T05:00:00.000Z",
      timeZone: "Asia/Ho_Chi_Minh",
    },
    summary: "NHAP MON DIEN TOAN (TH)",
  },
  {
    description: "lớp: 20221_CO1006_L23",
    end: {
      dateTime: "2022-11-17T09:50:00.000Z",
      timeZone: "Asia/Ho_Chi_Minh",
    },
    location: "HANGOUT_TUONGTAC",
    recurrence: [
      "RRULE:FREQ=WEEKLY;UNTIL=20221124T050000Z",
    ],
    start: {
      dateTime: "2022-11-17T05:00:00.000Z",
      timeZone: "Asia/Ho_Chi_Minh",
    },
    summary: "NHAP MON DIEN TOAN (TH)",
  },
  {
    description: "lớp: 20221_CO1006_L11",
    end: {
      dateTime: "2022-12-08T09:50:00.000Z",
      timeZone: "Asia/Ho_Chi_Minh",
    },
    location: "HANGOUT_TUONGTAC",
    recurrence: [],
    start: {
      dateTime: "2022-12-08T05:00:00.000Z",
      timeZone: "Asia/Ho_Chi_Minh",
    },
    summary: "NHAP MON DIEN TOAN (TH)",
  },
]
`;
